### PR TITLE
Use tsdx for build, test and lint

### PR DESCRIPTION
**Pull request by sveisvei**
Spawned as result of https://github.com/soywod/react-pin-field/issues/29

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -23,23 +23,21 @@
...
   },
-  "main": "lib/pin-field.min.js",
-  "types": "lib/pin-field.d.ts",
+  "main": "dist/index.js",
+  "typings": "dist/index.d.ts",
   "files": [
-    "lib/pin-field.min.js",
-    "lib/pin-field.d.ts"
+    "dist",
+    "src"
```
Comment (soywod) [inline package.json]: Is this line necessary? It means that all the sources are packed with releases, which increases the size of the lib. Maybe `dist` is enough. What do you think?

File: tsconfig.json [hunk 2]
```diff
@@ -1,10 +1,36 @@
...
+    "noImplicitReturns": true,
+    "noFallthroughCasesInSwitch": true,
+    // noUnused* overlap with @typescript-eslint/no-unused-vars, can disable if duplicative
+    "noUnusedLocals": true,
+    "noUnusedParameters": true,
+    "target": "esnext",
+    // use Node's module resolution algorithm, instead of the legacy TS one
+    "moduleResolution": "node",
+    // transpile JSX to React.createElement
+    "jsx": "react",
```
Comment (soywod) [inline tsconfig.json]: This can be removed since we do not use JSX/TSX.

File: tsconfig.json [hunk 3]
```diff
@@ -1,10 +1,36 @@
...
-    "target": "ES6"
-  },
-  "files": ["lib/pin-field.ts"]
+    // linter checks for common issues
+    "noImplicitReturns": true,
+    "noFallthroughCasesInSwitch": true,
+    // noUnused* overlap with @typescript-eslint/no-unused-vars, can disable if duplicative
+    "noUnusedLocals": true,
+    "noUnusedParameters": true,
+    "target": "esnext",
```
Comment (soywod) [inline tsconfig.json]: Sure about this one? Not all the browsers are understanding `ESNext` right now, I would stick with `ES6` (see https://www.typescriptlang.org/tsconfig#target):

> Modern browsers support all ES6 features, so ES6 is a good choice.

File: tsconfig.json [hunk 4]
```diff
@@ -1,10 +1,36 @@
...
+    "sourceMap": true,
+    // match output dir to input dir. e.g. dist/index instead of dist/src/index
+    "rootDir": "./src",
+    // stricter type-checking for stronger correctness. Recommended by TS
     "strict": true,
-    "target": "ES6"
-  },
-  "files": ["lib/pin-field.ts"]
+    // linter checks for common issues
+    "noImplicitReturns": true,
```
Comment (soywod) [inline tsconfig.json:12]: What are your motivations for this one?

File: tsconfig.json [hunk 5]
```diff
@@ -1,10 +1,36 @@
...
+    "noFallthroughCasesInSwitch": true,
```
Comment (soywod) [inline tsconfig.json]: What are your motivations for this one?

**Review by soywod — CHANGES_REQUESTED** [review]

File: tsconfig.json (hunk 5 above)
Comment (sveisvei) [reply tsconfig.json]: Good you are taking this on for your needs, like discussed. I did just copy the basic tsconfig from tsdx - and did minimal changes for it to look ok.

cheers

**Comment by soywod** [comment]
Thank you for your help @sveisvei, I completed your improvement and I just released it :tada:
I will take care of the React wrapper now (https://github.com/soywod/react-pin-field/pull/30).

**Pull request merged**